### PR TITLE
Fix for bug PAL-67.

### PR DESCRIPTION
**Pull request by saltshop-boy**
TIcket: https://dataworld.atlassian.net/browse/PAL-67

* Issue is seen when there are multiple  Rest API calls simultaneously to get the query result. (~10 or more calls)
* Caching logic is present to reduce these calls but it is not working. Every time script is called a new process
*  Google app script has cache service but it has 100KBytes size limitation. The query result can be more than 100KBytes. So response cannot be stored in the cache.
* Used Lock Service to make calls sequential. This is a user-level lock.

### CODE REVIEW

File: src/rest.js [hunk 1]
```diff
@@ -14,53 +14,50 @@
...
+        muteHttpExceptions: true
+    };
 
-        var response = UrlFetchApp.fetch(getSqlEndpoint(datasetKey), params);
-        if (response.getResponseCode() === 200) {
-            cachedResponse = response;
-            cachedResponseLastRefresh = lastRefresh;
+    var success = false;
+    var lock = LockService.getUserLock();
+    success = lock.tryLock(60000);
```
Comment (avaghasia-dw) [inline src/rest.js:32]: why `60000` for timeout?

File: src/rest.js [hunk 2]
```diff
@@ -14,53 +14,50 @@
...
+    
+    var response = UrlFetchApp.fetch(getSqlEndpoint(datasetKey), params);
```
Comment (avaghasia-dw) [inline src/rest.js:34]: if success is false, won't this still get executed?

Comment (saltshop-boy) [reply src/rest.js:34]: Yes, I will add code to throw an error
` "Server Busy please try after some time"`

File: src/rest.js (hunk 1 above)
Comment (saltshop-boy) [reply src/rest.js:32]: Maximum query result size from data.world connector is 25MB. Rest API will take ~2 seconds to query the result with 25MB data. I  assumed that there will not be more than 30 REST API call simultaneously and there may not always 25 MB of data in the query result. .  (Number 30 is a wild guess ).

 if you feel timeout 60 seconds is less or logic needs to change, please let me know.

Comment (saltshop-boy) [reply src/rest.js]: I have increased the timeout to 3minutes

**Review by avaghasia-dw — APPROVED** [review]
This seems good, is there a way to test?

**Comment by saltshop-boy** [comment]
Manually I have tested this using the private connector.   Using private I initially reproduce the issue and later verified that fix is working.

**Pull request merged**